src/Stratis.Bitcoin/EventBus/InMemoryEventBus.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Stratis.Bitcoin.EventBus
+{
+    public class InMemoryEventBus : IEventBus
+    {
+        /// <summary>
+        /// The subscriptions stored by EventType
+        /// </summary>
+        private readonly Dictionary<Type, List<ISubscription>> subscriptions;
+
+        /// <summary>
+        /// The subscriptions lock to prevent race condition during publishing
+        /// </summary>
+        private static readonly object SubscriptionsLock = new object();
+
+        public InMemoryEventBus()
+        {
+            this.subscriptions = new Dictionary<Type, List<ISubscription>>();
+        }
+
+        /// <inheritdoc />
+        public SubscriptionToken Subscribe<TEventBase>(Action<TEventBase> action) where TEventBase : EventBase
+        {
+            if (action == null)
+                throw new ArgumentNullException(nameof(action));
+
+            lock (SubscriptionsLock)
+            {
+                if (!this.subscriptions.ContainsKey(typeof(TEventBase)))
+                    this.subscriptions.Add(typeof(TEventBase), new List<ISubscription>());
+
+                var token = new SubscriptionToken(this, typeof(TEventBase));
+                this.subscriptions[typeof(TEventBase)].Add(new Subscription<TEventBase>(action, token));
+                return token;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Unsubscribe(SubscriptionToken token)
+        {
+            if (token == null)
+                throw new ArgumentNullException(nameof(token));
+
+            lock (SubscriptionsLock)
+            {
+                if (this.subscriptions.ContainsKey(token.EventType))
+                {
+                    var allSubscriptions = this.subscriptions[token.EventType];
+
+                    var subscriptionToRemove = allSubscriptions.FirstOrDefault(sub => sub.SubscriptionToken.Token == token.Token);
+                    if (subscriptionToRemove != null)
```

Comment:
Should we remove the type if no subscribers?

Reply:
It's unlikely that we remove a subscriber of a specific event type and never register again

Anyway this is a base implementation, an improvement I wanted already to implement but skipped at the last moment to not add too many things at first, is having fine subscription system (like I could put a condition that I want to receive only events of type, let say BlockDownloaded but only for blocks with an even height) and this mean that in that case having null doesn't mean that the subscriber of that events are 0 but just that the condition isn't met

Reply:
Over engineering in my opinion, just put any condituon you have on the receiver.

Reply:
this is why I didn't included it, I was sure you'd have said something about overengineering :D
anyway this is how many of the eventbus works, they allow you to cherrypick the subscription based on your needs, and if you think about it, it keeps the code cleaner because what you are interested in is stated in the subscription itself and not handled manually inside the event handler